bve-render/src/lib.rs

```diff
@@ -0,0 +1,378 @@
+                    ty: BindingType::UniformBuffer { dynamic: false },
+                },
+                BindGroupLayoutEntry {
+                    binding: 1,
+                    visibility: ShaderStage::FRAGMENT,
+                    ty: BindingType::SampledTexture {
+                        multisampled: false,
+                        component_type: TextureComponentType::Uint,
+                        dimension: TextureViewDimension::D2,
+                    },
+                },
+                BindGroupLayoutEntry {
+                    binding: 2,
+                    visibility: ShaderStage::FRAGMENT,
+                    ty: BindingType::Sampler { comparison: false },
+                },
+            ],
+        });
+        let pipeline_layout = device.create_pipeline_layout(&PipelineLayoutDescriptor {
+            bind_group_layouts: &[&bind_group_layout],
+        });
+
+        let opaque_pipeline = render::create_pipeline(
+            &device,
+            &pipeline_layout,
+            &vs_module,
+            &fs_module,
+            render::PipelineType::Normal,
+            samples,
+        );
+        let alpha_pipeline = render::create_pipeline(
+            &device,
+            &pipeline_layout,
+            &vs_module,
+            &fs_module,
+            render::PipelineType::Alpha,
+            samples,
+        );
+
+        let mip_creator = compute::MipmapCompute::new(&device);
+
+        // We need to do a couple operations on the whole pile first
+        let mut renderer = Self {
+            objects: IndexMap::new(),
+            object_handle_count: 0,
+
+            textures: IndexMap::new(),
+            texture_handle_count: 0,
+
+            camera: camera::Camera {
+                location: Vector3::new(-6.0, 0.0, 0.0),
+                pitch: 0.0,
+                yaw: 0.0,
+            },
+            screen_size,
+            samples,
+
+            surface,
+            device,
+            queue,
+            swapchain,
+            framebuffer,
+            depth_buffer,
+            opaque_pipeline,
+            alpha_pipeline,
+            pipeline_layout,
+            bind_group_layout,
+            sampler,
+
+            vert_shader: vs_module,
+            frag_shader: fs_module,
+
+            mip_creator,
+
+            command_buffers: Vec::new(),
+        };
+
+        // Default texture is texture handle zero, immediately discard the handle, never to be seen again
+        renderer.add_texture(&RgbaImage::from_raw(1, 1, vec![0xff, 0xff, 0xff, 0xff]).expect("Invalid Image"));
+
+        renderer
+    }
+
+    pub fn set_location(&mut self, object::ObjectHandle(handle): &object::ObjectHandle, location: Vector3<f32>) {
+        let object: &mut object::Object = &mut self.objects[handle];
+
+        object.location = location;
+    }
+
+    pub fn resize(&mut self, screen_size: PhysicalSize<u32>, samples: render::MSAASetting) {
```

Comment:
You should have a dedicated way to change MSAA instead of bundling it with screen resize. Those are separate concerns

Reply:
Silly you answered your own question; I agree.

Reply:
I looked at the method name, saw "resize", and I was like "why do you need to recompile PSOs when you resize"

Then I looked at the method parameters and I was like "MSAA does not go here"